### PR TITLE
Docstrings and readme

### PR DESCRIPTION
**Pull request by maxtaylordavies**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,48 +1,162 @@
...
+To run it with default arguments, simply execute 
+```bash
+python scripts/train_agent_babyai.py
 ```
 
-## Generating language feedback for an existing BabyAI dataset
+The script also supports a number of arguments. The main ones are highlighted below; for a complete list refer to `src/utils/argparsing.py`.
+
+### Specifying a BabyAI level
+The `--level` argument determines which BabyAI level to use, defaulting to "GoToRedBallGrey". A list of levels can be found on [this page](https://minigrid.farama.org/environments/babyai/) under "Registered Configurations".
```
Comment (sabraaap) [inline README.md:33]: Can we make a note that the format for specifying the level is just the middle part of the config name? e.g., "BabyAI-GoToRedBallGrey-v0" -> "GoToRedBallGrey"

File: README.md [hunk 2]
```diff
@@ -1,48 +1,162 @@
...
 
-### Specifying the dataset
-Rather than providing the name of the dataset, simply provide the same argument values as when you created the dataset (or in case the dataset doesn't exist yet, it will be created now), e.g. ``--env-name "BabyAI-GoToRedBallGrey-v0" --num_episodes 1000 --include_timeout False --seed 42``.
+### Controlling feedback generation
+If conditioning on feedback (see previous section), you can determine which *type* of feedback is used by setting the `--feedback_mode` argument to one of 'all', 'rule', 'task', 'numerical' or 'random'. For an explanation of what these feedback types mean, see section 3 and appendix A.1 of the paper. If using random feedback, you can additionally specify the type of random feedback through `--random_mode`, which can be either 'english' (uses random words from the BabyAI vocabulary) or 'lorem' (uses lorem ipsum text).
```
Comment (sabraaap) [inline README.md:42]: 1. Maybe we could briefly explain what numerical does?
2. 'english' uses random (nonsensical) English sentences that _don't_ contain any of the words in the BabyAI vocabulary

File: README.md [hunk 3]
```diff
@@ -1,48 +1,162 @@
...
 ```
 
-## Generating language feedback for an existing BabyAI dataset
+The script also supports a number of arguments. The main ones are highlighted below; for a complete list refer to `src/utils/argparsing.py`.
+
+### Specifying a BabyAI level
+The `--level` argument determines which BabyAI level to use, defaulting to "GoToRedBallGrey". A list of levels can be found on [this page](https://minigrid.farama.org/environments/babyai/) under "Registered Configurations". 
+
+### Controlling dataset creation
+To avoid creating a new dataset everytime the script is run, you can use the `--load_existing_dataset` argument - this will check if a dataset matching all the relevant arguments exists, and load it if so. By default, datasets are generated using a random policy - to instead sample data from the training process of a PPO agent, set `--policy` to "ppo". For both random and PPO policies, `--num_steps` controls the number of environment steps to include in the training dataset. Finally, to exclude episodes that end as a result of timing out, set `--include_timeout` to false.
```
Comment (sabraaap) [inline README.md:36]: 1. We could include a link to the datasets used for the experiments in the NeurIPS GCRL submission https://www.dropbox.com/sh/b0bff46d4s230hr/AADJE4xu_Aliqd_mAv3kUdSda?dl=0
2. --num_steps is only used for ppo datasets - for random datasets, we specify the --num_train_seeds and --eps_per_seed

File: README.md [hunk 4]
```diff
@@ -1,48 +1,162 @@
...
+        return {"loss": loss}
+
+    def get_action(self, input: AgentInput, context=1, one_hot=False):
+        pass
 ```
 
-If the dataset specified doesn't exist yet locally, it will be generated when the ``_feedback.py`` script is run.
+Implementing any particular type of agent/model then mostly consists of overriding the `_forward`, `_compute_loss` and `get_action` methods.
+
+The file `agent/fdt/base.py` contains an `Agent` implementation for the feedback decision transformer. This is the file that contains almost all the relevant logic for the actual FDT model. The files `agent/fdt/atari.py` and `agent/fdt/minigrid.py` contain slightly specialised versions for use with Atari and Minigrid/BabyAI environments respectively. At the time of writing, the only difference is the details of the CNN used to embed image observations.
```
Comment (sabraaap) [inline README.md:92]: we're not actually including any Atari code for this version (remind me to tell you something about this)

File: README.md [hunk 5]
```diff
@@ -1,48 +1,162 @@
...
 
-### Where to find feedback
-Feedback is store in json files in the folder ``feedback_data``. You will find all feedback variations for a given dataset in the same json file, organised hierarchically by the feedback hyperparameters (type > mode > frequency). You can retrieved it using the appropriate combination of keys.
+### Dataset
+The `CustomDataset` class (`dataset/custom_dataset.py`) essentially represents an interface to a collection of `hdf5` files on disk, where each file contains some number of recorded steps. Each step consists of the elements (mission, observation, action, reward, feedback, truncated, terminated). A new `CustomDataset` can be created either from existing DQN replay data (e.g. for Atari), or from entirely new data, sampled from some specified policy (or randomly). Once a `CustomDataset` has been created and saved to disk (across some number of sharded files), we can sample data from it using the following three methods:
```
Comment (sabraaap) [inline README.md:95]: Same here - I got rid of the DQN replay data part

File: README.md [hunk 6]
```diff
@@ -1,48 +1,162 @@
 # feedback-DT
 
-## Creating a BabyAI dataset
-Currently, this can be run with three different parameters. The plan is to also add an RGB parameter.
-```src/_datsets.py --env_name --num_episodes --include_timeout```
+This repository houses all the code for our paper '*Is Feedback All You Need? Leveraging Natural Language Feedback in Goal-Conditioned Reinforcement Learning*', as well as ongoing work developing the project further. If you use any of the code for your sown work, please cite the paper as
```
Comment (sabraaap) [inline README.md:3]: sown -> own
Should we mention that the NeurIPS 2023 GCRL submission code can be found in the neurips_gcrl branch?

File: README.md [hunk 7]
```diff
@@ -1,48 +1,162 @@
...
+        model: Agent,
+        **kwargs,
+    ):
+        self._plot_loss(state)
+
+        if state.global_step % self.eval_step_interval == 0 and state.global_step > 0:
+            self._run_eval_and_plot(model, state, eval_type="efficiency", control=control)
+            model.save_checkpoint(self.output_dir, state.global_step)
+```
+What happens here is that we first update our plot of the training loss (this is saved to disk at each step), and then check whether the current step is a multiple of `self.eval_step_interval` - if it is, we call `_run_eval_and_plot`. This evaluates the performance of the current agent, and updates the eval plot(s) saved on disk. Specifically: we sample *n* instances of the environment; for each of those, we sample a trajectory from the agent being trained, and also from a random agent (which serves as a baseline). For both trajectories we record the return, the length, and whether it was successful. All three of these metrics are plotted throughout training against number of samples. To track any additional metrics, you can just add code to `_run_eval_and_plot`.
```
Comment (sabraaap) [inline README.md:162]: should we mention the --sample_interval argument here?

File: README.md
Comment (sabraaap) [inline README.md:1]: Looks great!

File: src/collator/collator.py
Comment (sabraaap) [inline src/collator/collator.py:1]: Nice work, thanks for tidying up the unused imports too

**Review by sabraaap — CHANGES_REQUESTED** [review]
Fantastic job! Thanks for also reformatting stuff 💪 There's only a few tweaks on the README (see comments), let me know when you thinks it's ready me to approve and merge

File: README.md (hunk 1 above)
Comment (maxtaylordavies) [reply README.md:33]: good idea

File: README.md (hunk 2 above)
Comment (maxtaylordavies) [reply README.md:42]: oops my bad 😅

File: README.md [hunk 8]
```diff
@@ -1,48 +1,162 @@
...
+        context_length=64,
+        scale=1,
+        gamma=0.99,
+        embedding_dim=128,
+        randomise_starts=False,
+        episode_dist="uniform",
+    ) -> None:
+```
+
+The collator also handles the responsibility of embedding the natural language data (mission strings, feedback strings). This lets us swap out different embeddings models without having to modify the agent architecture. Since we expect a lot of repetition in the values of these strings, we implement a memoised embedder: the collator maintains a local embeddings cache; each time we get a mission or feedback string to embed, we check if we already have an embedding for that cache in the string, and only run the embedding model if not.
```
[the remark this answers is not in the document]
Comment (maxtaylordavies) [reply README.md:114]: actually it is memoised https://en.wikipedia.org/wiki/Memoization (unless i'm using it wrong haha)

File: README.md (hunk 6 above)
Comment (maxtaylordavies) [reply README.md:3]: good idea

**Review by sabraaap — APPROVED** [review]
🤟
